gitcmd/gitcmd.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Package gitcmd contains utilities for common Git operations in a local repository, including
+// authentication with a remote repository.
+package gitcmd
+
+import (
+	"fmt"
+	"strings"
+)
+
+const githubPrefix = "https://github.com/"
+const azdoDncengPrefix = "https://dnceng@dev.azure.com/"
+
+// GitURLAuther manipulates a Git repository URL (GitHub, AzDO, ...) such that Git commands taking a
+// remote will work with the URL. This is intentionally vague: it could add an access token into the
+// URL, or it could simply make the URL compatible with environmental auth on the machine (SSH).
+type GitURLAuther interface {
+	// InsertAuth inserts authentication into the URL and returns it, or if the auther doesn't
+	// apply, returns the url without any modifications.
+	InsertAuth(url string) string
+}
+
+// GitHubSSHAuther turns an https-style GitHub URL into an SSH-style GitHub URL.
+type GitHubSSHAuther struct{}
+
+func (a GitHubSSHAuther) InsertAuth(url string) string {
+	if strings.HasPrefix(url, githubPrefix) {
+		return fmt.Sprintf("git@github.com:%v", strings.TrimPrefix(url, githubPrefix))
+	}
```

Comment:
I'm not sure if this module can contain go1.18 features, but if so you can use my favorite addition to the standard library in this release, [strings.Cut](https://pkg.go.dev/strings#Cut):

```suggestion
        if _, after, found := strings.Cut(url, githubPrefix); found {
        	return fmt.Sprintf("git@github.com:%v", after)
        }
```

This apply to other HasPrefix+TrimPrefix occurrences in this file.

Reply:
Hmm, go-infra is used in microsoft/go by https://github.com/microsoft/go/blob/microsoft/main/eng/_util/go.mod (each release branch). But `_util` won't use this particular package, and even if it did (or will), it's fine for `_util` to require a new version of Go. It's fine to use go1.18 to run the infra/wrappers that build go1.17.x.

This is indeed a really great new func--using it. 😄

Reply:
Noticed that `Cut` isn't quite `CutPrefix`, and to really be equivalent, need to check `before` too:

```go
	if before, after, found := strings.Cut(url, githubPrefix); before == "" && found {
```

This is pretty unlikely to happen for a significant chunk of a URL like this, but this made me think... I could add a really simple `cutPrefix` func to the package based on the `strings.Cut` API design, and implement it directly to avoid even the possibility of issues with standard library versioning:

```go
func cutPrefix(s, prefix string) (after string, found bool) {
	if strings.HasPrefix(s, prefix) {
		return s[len(prefix):], true
	}
	return s, false
}
```